File: LayoutTests/http/tests/navigation/resources/redirection-response.php
<?php
$status_code = $_GET['status'];

$uri = rtrim(dirname($_SERVER['PHP_SELF']), '/\\') . "/" . $_GET['target'];

switch ($status_code) {
    case 301:
        header("HTTP/1.1 301 Moved Permanently");
        header("Location: http://" . $_SERVER['HTTP_HOST'] . $uri);
        break;
    case 302:
        header("HTTP/1.1 302 Found");
        header("Location: http://" . $_SERVER['HTTP_HOST'] . $uri);
        break;
    case 303:
        header("HTTP/1.1 303 See Other");
        header("Location: http://" . $_SERVER['HTTP_HOST'] . $uri);
        break;
    case 307:
        header("HTTP/1.1 307 Temporary Redirect");
        header("Location: http://" . $_SERVER['HTTP_HOST'] . $uri);
        break;
    default:
        header("HTTP/1.1 500 Internal Server Error");
        echo "Unexpected status code ($status_code) received.";
}
?>

File: LayoutTests/http/tests/navigation/resources/redirection-response.py
#!/usr/bin/env python3

import os
import sys
from urllib.parse import parse_qs

query = parse_qs(os.environ.get('QUERY_STRING', ''), keep_blank_values=True)
status = int(query.get('status', [500])[0])
target = query.get('target', [''])[0]

uri = '{}/{}'.format('/'.join(os.environ.get('SCRIPT_URL', '').split('/')[:-1]), target)

sys.stdout.write('Content-Type: text/html\r\n')

host = os.environ.get('HTTP_HOST', '')
if query.get('host', [None])[0] is not None:
    host = query.get('host')[0]

if status in [301, 302, 303, 307, 308]:
    sys.stdout.write(
        'status: {}\r\n'
        'Location: http://{}{}\r\n\r\n'.format(status, host, uri)
    )
else:
    sys.stdout.write(
        'status: 500\r\n\r\n'
        'Unexpected status code ({}) received.'.format(status)
    )
